block-streamer/src/receiver.rs

```diff
@@ -0,0 +1,259 @@
+use anyhow::Result;
+use block_ship::{chunking::chunks_to_path, types::BlockWrapper};
+use cid::Cid;
+use iroh_unixfs::Block;
+use messages::{TransmissionChunk, TransmissionMessage};
+use std::{collections::BTreeMap, path::PathBuf};
+use tracing::info;
+
+pub struct Receiver {
+    // Map of cid to blocks
+    pub blocks: BTreeMap<Cid, Block>,
+    // Map of cid to root blocks
+    pub roots: BTreeMap<Cid, Block>,
+    // Map of cid_marker to cid
+    pub cids_to_build: BTreeMap<Vec<u8>, Vec<u8>>,
+    // Map of cid_marker to [chunk]
+    pub cid_chunks: BTreeMap<Vec<u8>, Vec<TransmissionChunk>>,
+}
+
+impl Receiver {
+    pub fn new() -> Receiver {
+        Receiver {
+            blocks: BTreeMap::new(),
+            roots: BTreeMap::new(),
+            cids_to_build: BTreeMap::new(),
+            cid_chunks: BTreeMap::new(),
+        }
+    }
+
+    pub fn handle_chunk_msg(&mut self, chunk: TransmissionChunk) -> Result<()> {
+        info!(
+            "Received chunk {} for CID: {:?}",
+            chunk.chunk_offset, chunk.cid_marker
+        );
+
+        self.cid_chunks
+            .entry(chunk.cid_marker.clone())
+            .and_modify(|vec| vec.push(chunk.clone()))
+            .or_insert(vec![chunk.clone()]);
+        Ok(())
+    }
+
+    pub fn handle_cid_msg(&mut self, cid: Vec<u8>) -> Result<()> {
+        let cid_marker = &cid[..10];
+        self.cids_to_build.entry(cid_marker.to_vec()).or_insert(cid);
+        Ok(())
+    }
+
+    pub fn attempt_block_assembly(&mut self) -> Result<()> {
+        for (cid_marker, cid) in &self.cids_to_build {
+            if let Some(cid_chunks) = &self.cid_chunks.get(cid_marker) {
+                if let Ok(wrapper) = BlockWrapper::from_chunks(cid, cid_chunks) {
+                    let block = wrapper.to_block()?;
+                    if !block.links().is_empty() {
+                        info!(
+                            "Found root block {} with links {:?}",
+                            &block.cid(),
+                            &block.links()
+                        );
+                        self.roots.insert(*block.cid(), block);
+                    } else {
+                        info!(
+                            "Found child block {} with {} bytes",
+                            &block.cid(),
+                            block.data().len()
+                        );
+                        self.blocks.insert(*block.cid(), block.clone());
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+
+    pub async fn attempt_tree_assembly(&self) -> Result<()> {
```

Comment:
What is the purpose of the `attempt` here in the fn name?

Reply:
Currently the function walks the list of root blocks and _attempts_ to assemble the associated tree of blocks, by first checking if all child blocks are present, and then writing out to a file if "complete". It probably isn't named super well, and I think it is likely too general of a function. But it will be going away soon once we have persistent storage and the ability to request the export of a tree, so I may just add a comment and leave it be for now.